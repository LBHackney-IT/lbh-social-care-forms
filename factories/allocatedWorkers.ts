import { Factory } from 'fishery';

import { Allocation } from 'types';

export const allocationFactory = Factory.define<Allocation>(({ sequence }) => ({
  id: sequence,
  allocatedWorkerTeam: 'Safeguarding and Reviewing Team',
  allocatedWorker: 'Officer Name',
  allocationEndDate: '2019-03-28 00:00:00',
  allocationStartDate: '2019-03-28 00:00:00',
  workerType: 'Consultant Social Worker',
  personId: 1234,
  caseStatus: 'Open',
  personName: 'foo',
  personAddress: 'the address',
  personDateOfBirth: '2020-03-20',
  ragRating: 'amber',
}));

export const mockedAllocation = allocationFactory.build();
export const mockedAllocations = [allocationFactory.build()];
