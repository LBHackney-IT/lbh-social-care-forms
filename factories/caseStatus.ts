import { Factory } from 'fishery';
import {
  CaseStatus,
  CaseStatusFields,
  AddCaseStatusFormData,
  CaseStatusAnswerDisplay,
} from 'types';

export const mockedCaseStatusFactory = Factory.define<CaseStatus>(
  ({ sequence }) => ({
    id: sequence,
    type: 'CIN',
    answers: [mockedStatusField.build()],
    startDate: '2021-01-01T02:00:00Z',
    endDate: '2021-12-01T02:00:00Z',
    notes: 'a lot to note',
  })
);

export const mockedCaseStatusAddRequest = Factory.define<AddCaseStatusFormData>(
  ({ sequence }) => ({
    personId: sequence,
    type: 'CIN',
    fields: [{ name: 'test', selected: 'test' }],
    startDate: '2021-01-01T02:00:00Z',
    endDate: '2021-12-01T02:00:00Z',
    notes: 'a lot to note',
    createdby: 'test@testington.com',
  })
);

export const mockedStatusField = Factory.define<CaseStatusFields>(() => ({
  option: 'placementType',
  value: 'K1',
  startDate: '2020-08-01',
  createdAt: '2021-10-11T10:54:32Z',
}));

export const mockedCaseStatusAnswers = Factory.define<CaseStatusAnswerDisplay>(
  () => ({
    startDate: '2021-09-10',
    endDate: '2021-10-10',
    status: [mockedStatusField.build()],
  })
);
