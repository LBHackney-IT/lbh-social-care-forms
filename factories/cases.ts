import { Factory } from 'fishery';

import { Case } from 'types';

export const caseFactory = Factory.define<Case>(({ sequence }) => ({
  recordId: sequence.toString(),
  personId: 123,
  dateOfEvent: '25/10/2020 13:49:43',
  formName: 'foorm',
  officerEmail: 'Fname.Lname@hackney.gov.uk',
  caseFormData: {
    note: 'I am the note',
    first_name: 'Foo',
    last_name: 'Bar',
    timestamp: '11/03/2021 16:51:56',
    mosaic_id: 123,
    form_name_overall: 'foo',
    worker_email: 'worker@hackney.gov.uk',
    form_name: 'Foo bar',
    context_flag: 'A',
    date_of_event: '25/10/2020 13:49:43',
  },
  deleted: false,
  isImported: false,
}));

export const mockedNote = caseFactory.build({ caseFormUrl: 'https://foo.bar' });
export const mockedAllocationNote = caseFactory.build({
  formName: 'Worker allocated',
  caseFormData: {
    allocation_id: 321,
    created_by: 'foo@hackney.gov.uk',
    form_name_overall: 'API_Allocation',
    form_name: 'Worker allocated',
  },
});
export const mockedDeallocationNote = caseFactory.build({
  formName: 'Worker allocated',
  caseFormData: {
    allocation_id: 321,
    created_by: 'foo@hackney.gov.uk',
    deallocation_reason: 'deallocated because',
    form_name_overall: 'API_Deallocation',
    form_name: 'Worker deallocated',
  },
});
export const mockedCaseNote = caseFactory.build({
  caseFormData: {
    form_name_overall: 'ASC_case_note',
    case_note_title: 'i am a case title',
  },
});
export const mockedDeletedCaseNote = caseFactory.build({
  caseFormData: {
    form_name_overall: 'ASC_case_note',
    case_note_title: 'i am a case title',
  },
  formName: 'adult-case-note',
  deleted: true,
});
export const mockedWarningNoteCase = caseFactory.build({
  formName: 'Warning Note',
  caseFormData: {
    form_name_overall: 'API_WarningNote',
    warning_note_id: 456,
  },
});
