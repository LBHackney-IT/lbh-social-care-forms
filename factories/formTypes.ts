import { Factory } from 'fishery';
import { FormValue, FormStatusValue } from 'types';

export const mockedFormValueFactory = Factory.define<FormValue>(() => ({
  name: 'test',
  description: 'description',
  options: [mockedFormStatusValueFactory.build()],
}));

export const mockedFormStatusValueFactory = Factory.define<FormStatusValue>(
  () => ({
    name: 'N1',
    description: 'Noo 1',
  })
);

export const mockedFormValue = [mockedFormValueFactory.build()];
