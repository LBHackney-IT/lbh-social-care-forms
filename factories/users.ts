import { Factory } from 'fishery';

import { User } from 'types';

export const userFactory = Factory.define<User>(() => ({
  name: 'foo',
  hasAdminPermissions: true,
  hasChildrenPermissions: true,
  hasAdultPermissions: true,
  hasDevPermissions: false,
  isAuditable: false,
  email: 'foo@bar.com',
  permissionFlag: 'A',
  isAuthorised: true,
  isInWorkflowsPilot: false,
  isInSafeguardingReviewing: false,
  isInPlacementManagementUnit: false,
}));

export const mockedOnlyAdultUser = userFactory.build({
  hasAdminPermissions: false,
  hasChildrenPermissions: false,
  hasDevPermissions: false,
});

export const mockedOnlyChildUser = userFactory.build({
  hasAdminPermissions: false,
  hasAdultPermissions: false,
  hasDevPermissions: false,
});

export const mockedAdminUser = userFactory.build({
  hasDevPermissions: false,
});

export const mockedUser = userFactory.build();

export const mockedUserInWorkflowsPilot = userFactory.build({
  isInWorkflowsPilot: true,
});
