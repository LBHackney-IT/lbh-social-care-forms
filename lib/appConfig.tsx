import React from 'react';
import { useContext } from 'react';

type AppConfig = {
  [identifier: string]: string | number;
};

const AppConfigContext = React.createContext<AppConfig | undefined>(undefined);

export const AppConfigProvider: React.FC<{ appConfig: AppConfig }> = ({
  appConfig,
  children,
}) => {
  return (
    <AppConfigContext.Provider value={appConfig}>
      {children}
    </AppConfigContext.Provider>
  );
};

export const useAppConfig: () => {
  getConfigValue: (identifier: string) => string | number;
} = () => {
  const appConfig = useContext(AppConfigContext);

  if (appConfig === undefined)
    throw new Error(
      'A <AppConfigProvider /> must be provided as a parent of this component'
    );

  return {
    getConfigValue: (identifier) => {
      const configValue = appConfig[identifier];

      if (configValue === undefined)
        throw new Error(`A value for ${identifier} is not defined`);

      return configValue;
    },
  };
};
