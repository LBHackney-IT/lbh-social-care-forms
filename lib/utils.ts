import {
  Form,
  Step,
  Field,
  TimetableAnswer,
} from '../data/flexibleForms/forms.types';
import { Resident } from 'types';

export interface Theme {
  name: string;
  steps: Step[];
}

type InitialValue = null | string | string[] | InitialValues[] | unknown;

export interface InitialValues {
  [key: string]: InitialValue;
}

/** Take a form's steps and organise them by theme for displaying in a task list */
export const groupByTheme = (form: Form): Theme[] =>
  form.steps.reduce((groups: Theme[], step) => {
    const name = step.theme;
    if (name && !groups.find((group) => group.name === name)) {
      groups.push({
        name: name,
        steps: [],
      });
    }
    const group = groups.find((group) => group.name === name);
    group?.steps.push(step);
    return groups;
  }, []);

/** Shorten a long string to a given number of words for displaying in previews */
export const truncate = (str: string, noWords: number): string => {
  if (str.split(' ').length > noWords) {
    return str.split(' ').splice(0, noWords).join(' ') + '...';
  } else {
    return str;
  }
};

const initiallyNull = new Set(['file']);
const initiallyFirstChoice = new Set(['select']);
const initiallyArray = new Set(['checkboxes', 'repeater', 'tags']);

export const days: {
  [key: string]: string;
} = {
  Mon: 'Monday',
  Tue: 'Tuesday',
  Wed: 'Wednesday',
  Thu: 'Thursday',
  Fri: 'Friday',
  Sat: 'Saturday',
  Sun: 'Sunday',
  'Any day': 'Any day',
};

export const times = ['Morning', 'Afternoon', 'Evening', 'Night', 'Any time'];

const generateInitialTimetableValues = (): TimetableAnswer => {
  const initialTimetableValues: TimetableAnswer = {};
  Object.keys(days).map((day) => {
    initialTimetableValues[day] = {};
    times.map((time) => (initialTimetableValues[day][time] = ''));
  });
  return initialTimetableValues;
};

/** Generate flexible initial values for a flexible schema */
export const generateInitialValues = (
  fields: Field[],
  person?: Resident
): InitialValues => {
  const initialValues: InitialValues = {};

  fields.map((field) => {
    if (field.type === 'repeaterGroup') {
      if (field.hiddenRepeater) {
        initialValues[field.id] = [];
      } else {
        initialValues[field.id] = [
          generateInitialValues(field.subfields || [], person),
        ];
      }
    } else if (field.type === 'timetable') {
      initialValues[field.id] = generateInitialTimetableValues();
      initialValues[`${field.id} total hours`] = '';
    } else if (field.type === 'datetime') {
      initialValues[field.id] = field.default || [];
    } else if (initiallyArray.has(field.type)) {
      initialValues[field.id] = [];
    } else if (initiallyNull.has(field.type)) {
      initialValues[field.id] = null;
    } else if (initiallyFirstChoice.has(field.type) && !field.default) {
      initialValues[field.id] = String(
        (person && field.prefill && person[field.prefill]) ||
          (field.choices && field.choices[0].value)
      );
    } else if (field.type === 'select' && field.default) {
      initialValues[field.id] = field.default;
    } else {
      initialValues[field.id] = String(
        (person && field.prefill && person[field.prefill]) ||
          field.default ||
          ''
      );
    }
  });
  return initialValues;
};

/** Push an element into an array, without duplicates */
export const pushUnique = <T>(array: T[], newElement: T): T[] =>
  Array.from(new Set(array).add(newElement));

/** Take the values of a timetable question and get the sum total hours */
export const getTotalHours = (values: TimetableAnswer): number => {
  let total = 0;
  if (typeof values === 'object') {
    total = Object.values(values)?.reduce(
      (sum, day) =>
        sum +
        Object.values(day)?.reduce((sum, time) => sum + (Number(time) || 0), 0),
      0
    );
  }
  if (typeof total === 'number') return total;
  return 0;
};
