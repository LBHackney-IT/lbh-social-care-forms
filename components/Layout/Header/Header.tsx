import React from 'react';
import Link from 'next/link';

import { useAuth } from 'components/UserContext/UserContext';
import { getUserType } from 'utils/user';
import Logo from './Logo';
import { canUserManageWorkers } from '../../../lib/permissions';
import { useAppConfig } from 'lib/appConfig';

const HeaderComponent = ({
  serviceName,
}: {
  serviceName: string;
}): React.ReactElement => {
  const { user } = useAuth();
  const { getConfigValue } = useAppConfig();

  return (
    <header className="lbh-header ">
      <div className="lbh-header__main">
        <div className="lbh-container lbh-header__wrapper">
          <div className="lbh-header__title">
            <a href="/" className="lbh-header__title-link">
              <Logo />
              <span className="lbh-header__logo-text"> Hackney </span>
              <span className="lbh-header__service-name">{serviceName}</span>
              {user && (
                <span className="govuk-tag lbh-tag lbh-tag--green">
                  {getUserType(user)}
                </span>
              )}
            </a>
          </div>

          {user && (
            <nav className="lbh-header__links" aria-label="Navigation menu">
              {canUserManageWorkers(user) && (
                <Link href="/workers">
                  <a className="govuk-header__link">Manage workers</a>
                </Link>
              )}
              <Link href="/">
                <a className="govuk-header__link">My work</a>
              </Link>
              <Link href={getConfigValue('workflowsPilotUrl') as string}>
                <a className="govuk-header__link">Workflows</a>
              </Link>
              <Link href="/teams">
                <a className="govuk-header__link">Teams</a>
              </Link>
              <Link href="/search">
                <a className="govuk-header__link">Search</a>
              </Link>
              <Link href="/logout">
                <a className="govuk-header__link">Sign out</a>
              </Link>
            </nav>
          )}
        </div>
      </div>
    </header>
  );
};

export default HeaderComponent;
