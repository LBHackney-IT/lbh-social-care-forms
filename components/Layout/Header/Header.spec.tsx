import { render, screen } from '@testing-library/react';

import Header from './Header';
import { UserContext } from 'components/UserContext/UserContext';
import { mockedUser, userFactory } from 'factories/users';
import * as saveData from 'utils/saveData';
import { AppConfigProvider } from 'lib/appConfig';

let mockedUseRouter = { pathname: 'pathname' };

jest.mock('next/router', () => ({
  asPath: 'path',
  useRouter: () => mockedUseRouter,
}));

jest.mock('./Logo', () => () => 'MockedLogo');

describe('Header component', () => {
  const props = {
    serviceName: 'Foo',
  };
  it('should render header links with form in progress link', () => {
    jest.spyOn(saveData, 'getData').mockImplementationOnce(() => ({
      first: {
        data: { id: '123' },
        formPath: '/form/foo-bar/',
        timestamp: '22/12/2020',
        title: 'Foo Bar',
        step: 'foo',
        deleteForm: jest.fn(),
      },
    }));
    const { getByText } = render(
      <AppConfigProvider
        appConfig={{ workflowsPilotUrl: 'http://example.com' }}
      >
        <UserContext.Provider
          value={{
            user: mockedUser,
          }}
        >
          <Header {...props} />
        </UserContext.Provider>
      </AppConfigProvider>
    );

    expect(getByText('Search')).toBeInTheDocument();
    expect(getByText('Workflows')).toBeInTheDocument();
    expect(getByText('Sign out')).toBeInTheDocument();
  });

  it('should render service name but no header links', () => {
    const { getByText, queryByText } = render(
      <AppConfigProvider
        appConfig={{ workflowsPilotUrl: 'http://example.com' }}
      >
        <UserContext.Provider
          value={{
            user: undefined,
          }}
        >
          <Header {...props} />
        </UserContext.Provider>
      </AppConfigProvider>
    );
    expect(getByText('Foo')).toBeInTheDocument();
    expect(queryByText('Search')).not.toBeInTheDocument();
    expect(queryByText('Workflows')).not.toBeInTheDocument();
    expect(queryByText('Sign out')).not.toBeInTheDocument();
  });

  it('should set active the correct link', () => {
    mockedUseRouter = {
      pathname: '/my-records',
    };
    const { asFragment } = render(
      <AppConfigProvider
        appConfig={{ workflowsPilotUrl: 'http://example.com' }}
      >
        <UserContext.Provider
          value={{
            user: mockedUser,
          }}
        >
          <Header {...props} />
        </UserContext.Provider>
      </AppConfigProvider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should show the "Manage workers" navigation item for admins', () => {
    render(
      <AppConfigProvider
        appConfig={{ workflowsPilotUrl: 'http://example.com' }}
      >
        <UserContext.Provider
          value={{
            user: userFactory.build({
              hasAdminPermissions: true,
              hasDevPermissions: false,
            }),
          }}
        >
          <Header {...props} />
        </UserContext.Provider>
      </AppConfigProvider>
    );

    screen.getByText('Manage workers');
  });

  it('should show the "Manage workers" navigation item for developers', () => {
    render(
      <AppConfigProvider
        appConfig={{ workflowsPilotUrl: 'http://example.com' }}
      >
        <UserContext.Provider
          value={{
            user: userFactory.build({
              hasAdminPermissions: false,
              hasDevPermissions: true,
            }),
          }}
        >
          <Header {...props} />
        </UserContext.Provider>
      </AppConfigProvider>
    );

    screen.getByText('Manage workers');
  });

  it('should not show the "Manage workers" navigation item for regular users', () => {
    render(
      <AppConfigProvider
        appConfig={{ workflowsPilotUrl: 'http://example.com' }}
      >
        <UserContext.Provider
          value={{
            user: userFactory.build({
              hasAdminPermissions: false,
              hasDevPermissions: false,
            }),
          }}
        >
          <Header {...props} />
        </UserContext.Provider>
      </AppConfigProvider>
    );

    expect(screen.queryByText('Manage workers')).toBeNull();
  });
});
