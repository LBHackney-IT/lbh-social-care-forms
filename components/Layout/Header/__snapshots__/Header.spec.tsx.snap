// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component should set active the correct link 1`] = `
<DocumentFragment>
  <header
    class="lbh-header "
  >
    <div
      class="lbh-header__main"
    >
      <div
        class="lbh-container lbh-header__wrapper"
      >
        <div
          class="lbh-header__title"
        >
          <a
            class="lbh-header__title-link"
            href="/"
          >
            MockedLogo
            <span
              class="lbh-header__logo-text"
            >
               Hackney 
            </span>
            <span
              class="lbh-header__service-name"
            >
              Foo
            </span>
            <span
              class="govuk-tag lbh-tag lbh-tag--green"
            >
              Admin
            </span>
          </a>
        </div>
        <nav
          aria-label="Navigation menu"
          class="lbh-header__links"
        >
          <a
            class="govuk-header__link"
            href="/workers"
          >
            Manage workers
          </a>
          <a
            class="govuk-header__link"
            href="/"
          >
            My work
          </a>
          <a
            class="govuk-header__link"
            href="http://example.com"
          >
            Workflows
          </a>
          <a
            class="govuk-header__link"
            href="/teams"
          >
            Teams
          </a>
          <a
            class="govuk-header__link"
            href="/search"
          >
            Search
          </a>
          <a
            class="govuk-header__link"
            href="/logout"
          >
            Sign out
          </a>
        </nav>
      </div>
    </div>
  </header>
</DocumentFragment>
`;
